### PR TITLE
Add ExemplarFilter interfaces

### PR DESCRIPTION
**Pull request by cijothomas**
Towards https://github.com/open-telemetry/opentelemetry-dotnet/issues/2527
Adds ExemplarFilter interface/default implnm.
Everything is internal, and is not connected to the Metrics SDK. This will be done post 1.4.0 stable release.

Ref: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exemplarfilter

### CODE REVIEW

File: src/OpenTelemetry/Metrics/Exemplar/AllExemplarFilter.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+using System.Collections.Generic;
+
+namespace OpenTelemetry.Metrics
+{
+    /// <summary>
+    /// The Exemplar Filter which always samples measurements./>
+    /// </summary>
+    internal sealed class AllExemplarFilter : IExemplarFilter
+    {
+        public bool ShouldSample(long value, ReadOnlySpan<KeyValuePair<string, object>> tags)
```
Comment (reyang) [inline src/OpenTelemetry/Metrics/Exemplar/AllExemplarFilter.cs:27]: The value has a `long` type, what's the current thinking of other types (e.g. double)?

File: src/OpenTelemetry/Metrics/Exemplar/NoneExemplarFilter.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenTelemetry.Metrics
```
Comment (reyang) [inline src/OpenTelemetry/Metrics/Exemplar/NoneExemplarFilter.cs:20]: minor: consider `namespace OpenTelemetr.Metrics;` (single lined approach to save indentation).

File: src/OpenTelemetry/DiagnosticSourceInstrumentation/ListenerHandler.cs [hunk 3]
```diff
@@ -42,9 +42,9 @@ public ListenerHandler(string sourceName)
         public virtual bool SupportsNullActivity { get; }
 
         /// <summary>
-        /// Method called for an event which does not have 'Start', 'Stop' or 'Exception' as suffix.
+        /// Method called for event.
         /// </summary>
-        /// <param name="name">Custom name.</param>
+        /// <param name="name">Event name.</param>
```
Comment (cijothomas) [inline src/OpenTelemetry/DiagnosticSourceInstrumentation/ListenerHandler.cs:47]: unrelated nit fix. happened to be in my local!

File: src/OpenTelemetry/Metrics/Exemplar/IExemplarFilter.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    /// Returns
+    /// <c>true</c> to indicate this measurement is eligible to become Exemplar
+    /// and will be given to an ExemplarReservoir.
+    /// Reservoir may further sample, so a true here does not mean that this
+    /// measurement will become an exemplar, it just means it'll be
+    /// eligible for being Exemplar.
+    /// <c>false</c> to indicate this measurement is not eligible to become Exemplar
+    /// and will not be given to the ExemplarReservoir.
+    /// </returns>
+    bool ShouldSample(long value, ReadOnlySpan<KeyValuePair<string, object>> tags);
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/Exemplar/IExemplarFilter.cs:43]: Do we need an explicit parameter for timestamp? When would this method be called with respect to when a metric is recorded?

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Exemplar/IExemplarFilter.cs:43]: spec says timestamp is needed. but we do not allow user to provide timestamp for metric APIs, and is always assumed to be the current time.. so the Filter can also take current timestamp, without we providing it. (have to check if this is bad for perf.. )

(Similar with context. we dont pass it. its implicit.

Comment (utpilla) [reply src/OpenTelemetry/Metrics/Exemplar/IExemplarFilter.cs:43]: No concerns with thread preemption here, right?

**Review by utpilla — APPROVED** [review]

File: src/OpenTelemetry/Metrics/Exemplar/IExemplarFilter.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenTelemetry.Metrics;
+
+/// <summary>
+/// The interface defining Exemplar Filter.
+/// </summary>
+internal interface IExemplarFilter
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/Exemplar/IExemplarFilter.cs:25]: Is "filter" coming from the spec? Asking because `IExemplarSampler`, `AlwaysOnExemplarSampler`, & `AlwaysOffExemplarSampler` might fit in better with what we have for tracing 🤷

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Exemplar/IExemplarFilter.cs:25]: https://github.com/open-telemetry/opentelemetry-specification/pull/2919 I am proposing fixes in spec in parallel.

File: src/OpenTelemetry/Metrics/Exemplar/AllExemplarFilter.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/Exemplar/AllExemplarFilter.cs]: Will we only ever have `long` & `double`? Just wondering if it should be `ShouldSample<T>(T value, ReadOnlySpan<KeyValuePair<string, object>> tags)`.

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Exemplar/AllExemplarFilter.cs]: at the earliest stage of callback from .NET runtime, the SDK converts everything to long or double.
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/MeterProviderSdk.cs#L262-L270

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Exemplar/AllExemplarFilter.cs]: in the future where we allow pluggable aggregators, they would also see long/double (per current design...)

**Comment by cijothomas** [comment]
Marking as draft. No need to merge now. I will update once the perf implications are fully understood with basic implementations.

**Comment by cijothomas** [comment]
I was planning to merge changes to main after 1.4.0 ship has sailed. Unfortunately 1.4.0 has not yet happened. I'll be creating a new branch for metrics, and maintain that temporarily until 1.4.0 is out. Starting next week, will be sending PRs to merge to the metrics branch. This is done to unblock Exemplar feature work.
